File: apps/api/app/editor_assist.py
"""編集者向けの危険度・引用メモ。法的断定はしない。"""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any

from app.jev_client import ACT_THRESHOLD

_EASY = re.compile(
    r"(CC0|Public Domain|パブリックドメイン)",
    re.IGNORECASE,
)
_CONDITIONAL = re.compile(
    r"(CC[-\s]?BY|Creative Commons|All rights reserved|クリエイティブ・コモンズ)",
    re.IGNORECASE,
)

_KIND_JA = {
    "own_work": "自作",
    "third_party": "他者",
    "web": "Web",
}


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def risk_level(source: dict[str, Any]) -> str:
    """easy | conditional | review。目安のみ。"""
    license_value = str(source.get("license") or "").strip()
    kind = str(source.get("material_kind") or "web")
    if license_value == "own_claim" or (license_value and _EASY.search(license_value)):
        return "easy"
    if license_value and _CONDITIONAL.search(license_value):
        return "conditional"
    if kind in ("web", "third_party"):
        jev = source.get("jev") or {}
        if isinstance(jev, dict):
            for key in ("license", "copyright"):
                score = jev.get(key)
                if score is not None and float(score) < ACT_THRESHOLD * 100 and license_value in (
                    "",
                    "unknown",
                ):
                    return "review"
        if not license_value or license_value == "unknown":
            return "review"
    if not license_value or license_value == "unknown":
        return "review"
    return "review"


def credit_memo(source: dict[str, Any], viewed_at: str = "") -> str:
    title = str(source.get("title") or "").strip() or "(無題)"
    url = str(source.get("url") or "").strip() or "(URLなし)"
    kind = str(source.get("material_kind") or "web")
    kind_label = _KIND_JA.get(kind, kind)
    license_value = str(source.get("license") or "unknown").strip() or "unknown"
    when = viewed_at.strip() or now_iso()
    return (
        f"タイトル: {title}\n"
        f"URL: {url}\n"
        f"種別: {kind_label}\n"
        f"ライセンス: {license_value}\n"
        f"取得日: {when}"
    )


def enrich_sources(sources: list[dict[str, Any]], viewed_at: str) -> list[dict[str, Any]]:
    enriched: list[dict[str, Any]] = []
    for item in sources:
        row = dict(item)
        # 自動検索の出典には license が無いことがある
        if "license" not in row:
            if row.get("material_kind") == "own_work":
                row["license"] = "own_claim"
            else:
                row["license"] = "unknown"
        row["risk_level"] = risk_level(row)
        row["credit_memo"] = credit_memo(row, viewed_at)
        enriched.append(row)
    return enriched


def enrich_turn(payload: dict[str, Any], object_label: str) -> dict[str, Any]:
    viewed = now_iso()
    sources = enrich_sources(list(payload.get("sources") or []), viewed)
    return {
        **payload,
        "object_label": object_label,
        "viewed_at": viewed,
        "sources": sources,
    }

File: apps/web/src/lib/editorAssist.ts
/** 編集者向けの危険度・引用メモ・書き出し。法的断定はしない。 */

import type { ChatTurn, Lookup, LookupSource, MaterialKind, Source } from "@/lib/api";

export type RiskLevel = "easy" | "conditional" | "review";

export type CheckAnswer = "yes" | "no" | "unknown";

export type SourceChecklist = {
  needs_credit: CheckAnswer;
  allow_modify: CheckAnswer;
  commercial: CheckAnswer;
  not_confused: CheckAnswer;
};

export const EMPTY_CHECKLIST: SourceChecklist = {
  needs_credit: "unknown",
  allow_modify: "unknown",
  commercial: "unknown",
  not_confused: "unknown",
};

const EASY = /(CC0|Public Domain|パブリックドメイン)/i;
const CONDITIONAL = /(CC[-\s]?BY|Creative Commons|All rights reserved|クリエイティブ・コモンズ)/i;

const KIND_JA: Record<MaterialKind, string> = {
  own_work: "自作",
  third_party: "他者",
  web: "Web",
};

export function riskLevel(source: {
  license?: string;
  material_kind?: MaterialKind;
  risk_level?: RiskLevel;
}): RiskLevel {
  if (source.risk_level === "easy" || source.risk_level === "conditional" || source.risk_level === "review") {
    return source.risk_level;
  }
  const license = (source.license || "").trim();
  if (license === "own_claim" || (license && EASY.test(license))) {
    return "easy";
  }
  if (license && CONDITIONAL.test(license)) {
    return "conditional";
  }
  return "review";
}

export function creditMemo(
  source: {
    title?: string;
    url?: string;
    material_kind?: MaterialKind;
    license?: string;
    credit_memo?: string;
  },
  viewedAt = "",
): string {
  if (source.credit_memo?.trim()) {
    return source.credit_memo.trim();
  }
  const title = source.title?.trim() || "(無題)";
  const url = source.url?.trim() || "(URLなし)";
  const kind = source.material_kind || "web";
  const license = source.license?.trim() || "unknown";
  const when = viewedAt.trim() || new Date().toISOString();
  return [
    `タイトル: ${title}`,
    `URL: ${url}`,
    `種別: ${KIND_JA[kind] || kind}`,
    `ライセンス: ${license}`,
    `取得日: ${when}`,
  ].join("\n");
}

export function splitByOwnership<T extends { material_kind?: MaterialKind }>(
  sources: T[],
): { own: T[]; other: T[] } {
  const own: T[] = [];
  const other: T[] = [];
  for (const item of sources) {
    if (item.material_kind === "own_work") {
      own.push(item);
    } else {
      other.push(item);
    }
  }
  return { own, other };
}

export function kindCounts(sources: { material_kind?: MaterialKind }[]): string {
  let own = 0;
  let other = 0;
  for (const item of sources) {
    if (item.material_kind === "own_work") {
      own += 1;
    } else {
      other += 1;
    }
  }
  return `自作${own} / 他者${other}`;
}

export function formatViewedAt(iso: string | undefined): string {
  if (!iso) {
    return "";
  }
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) {
    return iso;
  }
  return date.toLocaleString("ja-JP", { hour12: false });
}

export function isFavoriteCandidate(
  source: { url?: string; license?: string; material_kind?: MaterialKind; risk_level?: RiskLevel },
  checklist: SourceChecklist,
): boolean {
  if (riskLevel(source) !== "easy") {
    return false;
  }
  return (
    checklist.needs_credit !== "unknown" &&
    checklist.allow_modify !== "unknown" &&
    checklist.commercial !== "unknown" &&
    checklist.not_confused !== "unknown"
  );
}

export function buildExportMarkdown(
  history: Lookup[],
  chatHistory: ChatTurn[],
  checks: Record<string, SourceChecklist>,
  objectLabels: Record<string, string>,
  favorites: Record<string, true> = {},
  notes: Record<string, string> = {},
): string {
  const lines: string[] = ["# 制作ログ（出典下書き）", "", "> 目安です。法的な断定ではありません。", ""];

  const favoriteUrls = Object.keys(favorites).filter((url) => favorites[url]);
  if (favoriteUrls.length) {
    lines.push("## お気に入り", "");
    for (const url of favoriteUrls) {
      const note = (notes[url] || "").trim();
      lines.push(`- ${url}${note ? ` — ${note}` : ""}`);
    }
    lines.push("");
  }

  if (history.length) {
    lines.push("## 自動検索", "");
    for (const turn of history) {
      const label = turn.object_label || objectLabels[turn.object_id] || turn.object_id;
      lines.push(`### ${label} · ${formatViewedAt(turn.viewed_at)}`);
      lines.push("");
      lines.push(turn.summary || "");
      lines.push("");
      for (const source of turn.sources) {
        lines.push(..._sourceBlock(source, turn.viewed_at, checks, notes));
      }
    }
  }

  if (chatHistory.length) {
    lines.push("## チャット判断", "");
    for (const turn of chatHistory) {
      const label = turn.object_label || objectLabels[turn.object_id] || turn.object_id;
      lines.push(`### ${label} · ${formatViewedAt(turn.viewed_at)}`);
      lines.push("");
      lines.push(`質問: ${turn.message}`);
      lines.push("");
      lines.push(turn.answer || "");
      lines.push("");
      for (const source of turn.sources) {
        lines.push(..._sourceBlock(source, turn.viewed_at, checks, notes));
      }
    }
  }

  return `${lines.join("\n").trim()}\n`;
}

function _sourceBlock(
  source: LookupSource | Source,
  viewedAt: string | undefined,
  checks: Record<string, SourceChecklist>,
  notes: Record<string, string>,
): string[] {
  const key = source.url || source.title || "";
  const check = checks[key] || EMPTY_CHECKLIST;
  const license = "license" in source ? source.license : "unknown";
  const note = (notes[key] || "").trim();
  const block = [
    "```",
    creditMemo({ ...source, license }, viewedAt || ""),
    `危険度目安: ${riskLevel({ ...source, license })}`,
    `チェック: クレジット=${check.needs_credit} / 改変=${check.allow_modify} / 商用=${check.commercial} / 混同なし=${check.not_confused}`,
  ];
  if (note) {
    block.push(`メモ: ${note}`);
  }
  block.push("```", "");
  return block;
}

export async function copyText(text: string): Promise<boolean> {
  try {
    await navigator.clipboard.writeText(text);
    return true;
  } catch {
    return false;
  }
}

export function downloadText(filename: string, body: string): void {
  const blob = new Blob([body], { type: "text/plain;charset=utf-8" });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  anchor.click();
  URL.revokeObjectURL(url);
}
